Visualization/bokeh_holoviews_datashader.ipynb

```diff
@@ -0,0 +1,560 @@
+    "kdims = [('ra', 'RA(deg)'), ('dec', 'Dec(deg)')]\n",
+    "vdims = [('phot_g_mean_mag', 'G(mag)')]\n",
+    "ds = hv.Dataset(df, kdims, vdims)\n",
+    "ds"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "points = hv.Points(ds)\n",
+    "\n",
+    "#boundsxy = (0, 0, 0, 0)\n",
+    "boundsxy = (np.min(ds.data['ra']), np.min(ds.data['dec']), np.max(ds.data['ra']), np.max(ds.data['dec']))\n",
+    "box = streams.BoundsXY(source=points, bounds=boundsxy)\n",
+    "box_plot = hv.DynamicMap(lambda bounds: hv.Bounds(bounds), streams=[box])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# This function defines the custom callback functionality to update the linked histogram\n",
+    "def update_histogram(bounds=bounds):\n",
+    "    \n",
+    "    selection = (ds.data['ra'] > bounds[0]) & \\\n",
+    "                (ds.data['dec'] > bounds[1]) & \\\n",
+    "                (ds.data['ra'] < bounds[2]) & \\\n",
+    "                (ds.data['dec'] < bounds[3])\n",
+    "    \n",
+    "    selected_mag = ds.data.loc[selection]['phot_g_mean_mag']\n",
+    "    \n",
+    "    frequencies, edges = np.histogram(selected_mag)\n",
+    "    \n",
+    "    hist = hv.Histogram((np.log(frequencies), edges))\n",
+    "    return hist"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%output size=150\n",
+    "dmap = hv.DynamicMap(update_histogram, streams=[box])\n",
+    "datashade(points, cmap=bokeh.palettes.Viridis256) * box_plot + dmap"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Notice that when you select different regions of the left panel with the box select tool, the histogram on the right is updated."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 3: Images\n",
+    "\n",
+    "The next example demonstrates image visualization at the pixel level with datashader."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# First get a sensor image\n",
+    "import lsst.daf.persistence as dafPersist\n",
+    "\n",
+    "butler = dafPersist.Butler('/project/shared/data/Twinkles_subset/output_data_v2')\n",
+    "subset = butler.subset('calexp')\n",
+    "dataid = subset.cache[0]\n",
+    "image = butler.get('calexp', immediate=True, dataId=dataid)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%opts Image  [height=600 width=650]\n",
+    "%%opts Bounds (color='white')\n",
+    "#%%output size=200\n",
+    "\n",
+    "# Make a fake image just for testing purpose\n",
```

Comment:
Do you want to remove the fake image code that's commented out, or is it useful to keep it in?

Reply:
Probably more of a distraction than useful. I'll remove.